provider/openstack/provider.go

```diff
@@ -1424,11 +1425,63 @@ func (e *Environ) networksForInstance() ([]nova.ServerNetworks, error) {
 			return nil, errors.Trace(err)
 		}
 	} else {
-		logger.Debugf("using network id %q", networkId)
-		networks = append(networks, nova.ServerNetworks{NetworkId: networkId})
+		logger.Debugf("using network id %q", networkID)
+		networks = append(networks, nova.ServerNetworks{
+			NetworkId: networkID,
+		})
+	}
+
+	// Attempt to locate any subnet IDs for a passed in AZ.
+	availabilityZone := args.AvailabilityZone
+
+	var subnetIDsForZone []string
+	if args.Constraints.HasSpaces() {
+		var err error
+		subnetIDsForZone, err = corenetwork.FindSubnetIDsForAvailabilityZone(availabilityZone, args.SubnetsToZones)
+
+		switch {
+		case errors.IsNotFound(err):
+			return nil, errors.Trace(err)
+		case err != nil:
+			return nil, errors.Annotatef(err, "getting subnets for zone %q", availabilityZone)
+		}
+	}
+
+	// TODO (stickupkid): This is has commonalities with EC2 provider. We should
+	// work out if it's possible to extract into core/network
+
+	// If there are some subnet IDs from an AZ, attempt to put them on the
+	// networks.
+	// TODO (stickupkid): Currently this logic attempts to handle multiple
+	// networks, in fact in reality it actually handles only one server network.
+	// DefaultNetworks currently returns an empty slice and ResolveNetwork only
+	// returns one network. bug #1733266
+	var subnetID string
+	if num := len(subnetIDsForZone); num > 1 {
+		// Randomize the subnetIDsForZone in order to gain a better distributed
+		// spread over the zones.
+		// Note: this won't perform an even distribution unless a new random
+		// seed has been given for every initialization.
+		subnetID = subnetIDsForZone[rand.Intn(len(subnetIDsForZone))]
+		logger.Debugf("selected random subnet %q from all matching in zone %q", subnetID, availabilityZone)
+	} else if num == 1 {
+		subnetID = subnetIDsForZone[0]
+		logger.Debugf("selected subnet %q in zone %q", subnetID, availabilityZone)
+	}
+
+	// Set the subnetID on the network for all networks.
+	subnetNetworks := make([]nova.ServerNetworks, len(networks))
+	for k, network := range networks {
+		// If no networkId is found we need to skip them as we can't do anything
+		// with them.
+		if network.NetworkId == "" {
+			continue
+		}
+		network.FixedIp = subnetID
```

Comment:
Bummer. I just realised something.

The AZ map works OK for EC2 because it uses its own subnet IDs for provisioning. The docs for `FixedIp` say that it needs to be a CIDR.

Hopefully we can ask the provider what the CIDR is for the subnet ID. Otherwise we might have to change what we pass around from the provisioner.

Reply:
I dig in the docks to see if we can work this out...